examples/apps/windows/ClientControl.xaml.cpp

```diff
@@ -0,0 +1,215 @@
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "pch.h"
+#include <utility>
+#include <algorithm>
+#include "ClientControl.xaml.h"
+#include "Factory.h"
+
+using namespace Thread;
+
+using namespace Concurrency;
+using namespace Platform;
+using namespace Windows::ApplicationModel::Core;
+using namespace Windows::Foundation;
+using namespace Windows::Foundation::Collections;
+using namespace Windows::UI;
+using namespace Windows::UI::Xaml;
+using namespace Windows::UI::Xaml::Controls;
+using namespace Windows::UI::Xaml::Controls::Primitives;
+using namespace Windows::UI::Xaml::Data;
+using namespace Windows::UI::Xaml::Input;
+using namespace Windows::UI::Xaml::Media;
+using namespace Windows::UI::Xaml::Navigation;
+
+// The User Control item template is documented at http://go.microsoft.com/fwlink/?LinkId=234236
+
+std::atomic<int> ClientControl::_clientPort{ TalkConsts::DEF_CLIENT_PORT_INIT };
+
+ClientControl::ClientControl()
+{
+    InitializeComponent();
+
+    ServerPort->Text = DEF_SERVER_PORT.ToString();
+    auto clientPort = _clientPort.load();
+    ClientPort->Text = clientPort.ToString();
+}
+
+void
+ClientControl::Init(
+    IAsyncThreadNotify^  notify,
+    IMainPageUIElements^ mainPageUIElements)
+{
+    _notify = std::move(notify);
+    _mainPageUIElements = std::move(mainPageUIElements);
+}
+
+void
+ClientControl::ProtocolChanged(
+    Protocol protocol)
+{
+    _protocol = protocol;
+}
+
+void
+ClientControl::Connect_Click(
+    Object^          sender,
+    RoutedEventArgs^ e)
+{
+    try
+    {
+        auto clientArgs = ref new ClientArgs();
+
+        auto serverIP = ServerIP->Text;
+        if (serverIP->IsEmpty())
+        {
+            Exception::CreateException(E_INVALIDARG, "No Server IP");
+        }
+
+        // check valid chars of Ipv6 Address
+        if (std::all_of(serverIP->Data(), serverIP->Data() + serverIP->Length(),
+            [](wchar_t c)
+        {
+            if ((c >= '0' && c <= '9') || (c == ':') || (c >= 'a' && c <= 'z'))
```

Comment:
Section 4 of RFC 5952 reads:
> The recommendation in this section SHOULD be followed by systems when generating an address to be represented as text, but all implementations MUST accept and be able to handle any legitimate [RFC4291] format.

This seams to mean that when displaying/generating an address you need to use lower case, but must be able to take both upper and lower as input.